### PR TITLE
[WIP] Expanding pool

### PR DESCRIPTION
**Pull request by rgbkrk**
This adds an endpoint for expanding a pool size. Needs auth, I've stubbed it out for now.

In addition, this includes culling dangling containers within the heartbeat: containers which are neither routed nor in the pool. We have to assume they should be deleted since they're certainly not reachable by users but we also have no idea if another user has touched it. I may open this up in a separate PR.

### CODE REVIEW

File: Dockerfile [hunk 1]
```diff
@@ -12,4 +12,8 @@ ENV DOCKER_HOST unix://docker.sock
 
 RUN pip install -r requirements.txt
 
-CMD python orchestrate.py
+# Since old instructions list "python orchestrate.py" make sure we
+# direct to python3 underneath
+RUN ln -sf "$( which python3 )" "$( dirname $( which python3 ) )/python"
```
Comment (captainsafia) [inline Dockerfile]: Ha! Nice.

File: orchestrate.py [hunk 2]
```diff
@@ -156,6 +169,47 @@ def post(self):
...
+        if 'pool' in model and 'capacity' in model['pool']:
+            new_capacity = int(model['pool']['capacity'])
+            self.pool.capacity = new_capacity
+            yield self.pool.heartbeat()
+
+        diagnosis = self.pool.diagnosis()
+        yield diagnosis.observe()
+        routes = yield diagnosis._proxy_routes()
+
+        diagnosisDict = dict(
```
Comment (smashwilson) [inline orchestrate.py]: Nitpick: `diagnosis_dict` maybe? Mixing camel and snake case is :-1:.

File: orchestrate.py [hunk 3]
```diff
@@ -248,7 +302,8 @@ def main():
         (r"/(user/\w+)(?:/.*)?", LoadingHandler),
         (r"/api/stats/?", APIStatsHandler),
         (r"/stats/?", RedirectHandler, {"url": "/api/stats"}),
-        (r"/info/?", InfoHandler)
+        (r"/info/?", InfoHandler),
+        (r"/api/pool/1?", APIPoolHandler),
```
Comment (smashwilson) [inline orchestrate.py]: What's the `/1?` bit for?

File: orchestrate.py [hunk 4]
```diff
@@ -156,6 +169,47 @@ def post(self):
...
+        # TODO: Auth
+        # TODO: Determine payload format
+        model = self.get_json_body()
+        if model is None:
+            return self.write({'error': 'empty payload'})
+
+        if 'pool' in model and 'capacity' in model['pool']:
+            new_capacity = int(model['pool']['capacity'])
+            self.pool.capacity = new_capacity
+            yield self.pool.heartbeat()
```
Comment (smashwilson) [inline orchestrate.py]: There's probably a better (stricter) way to validate the JSON structure. I'd be surprised if there wasn't a library for it somewhere.

Also, I'd argue it should return a 400 for missing required properties. It's a bit weird to me that the `POST` degrades to "give me the current status" with no error if you post valid JSON to it.

File: orchestrate.py (hunk 2 above)
Comment (rgbkrk) [inline orchestrate.py]: I've been writing too much node and go, methinks.

File: orchestrate.py (hunk 3 above)
Comment (rgbkrk) [inline orchestrate.py]: I just realized this should be `/api/pools`. There's only ever one pool here, to be used by some higher level "make a new tmpnb/pool" via this API proposal: https://github.com/jupyter/enhancement-proposals/pull/5

File: orchestrate.py (hunk 2 above)
Comment (minrk) [inline orchestrate.py]: s/methinks/meThinks/

File: orchestrate.py (hunk 4 above)
Comment (rgbkrk) [inline orchestrate.py]: > There's probably a better (stricter) way to validate the JSON structure. I'd be surprised if there wasn't a library for it somewhere.

Yeah, we use `jsonschema` in the notebook server for notebooks.

This was loosey goosey while debugging and seeing what was going on with the spawn pool and diagnostics, I definitely plan to rip it out.

**Comment by rgbkrk** [comment]
Closing this for now until picked up again.

**Pull request closed without merging**